pkg/logs/input/docker/launcher.go

```diff
@@ -56,8 +59,22 @@ func NewLauncher(sources *config.LogSources, services *service.Services, pipelin
 		erroredContainerID: make(chan string),
 		lock:               &sync.Mutex{},
 	}
-	err := launcher.setup()
-	if err != nil {
+
+	var retryStrategy retry.Strategy
+	if shouldRetry {
+		retryStrategy = retry.RetryCount
+	} else {
+		retryStrategy = retry.OneTry
+	}
+	launcher.initRetry.SetupRetrier(&retry.Config{
+		Name:          "docker Launcher setup",
+		AttemptMethod: func() error { return launcher.setup() },
+		Strategy:      retryStrategy,
+		RetryCount:    math.MaxInt32,
```

Comment:
Should there be a maximum trial period after which the log agent stops waiting for the docker daemon to show up ?
Or should the log agent try to connect to docker forever ?

Reply:
`try to connect to docker forever` means if there is no docker, we will always look for one every 30 seconds? (would that be costly?)

Reply:
Yes. It means it will only try to connect to the unix domain socket `/var/run/docker.sock` and get a connection refused error once every ½ minute.

Reply:
Happy to see that we have an abstraction for retries, what's the strategy underneath ? Exponential ? Can we configure it ?

Reply:
Unfortunately, the current `retrier` implementation doesn’t implement exponential back-off.
So far, it has two simple strategies:
* **OneTry**: that will ensure that the `AttemptMethod` is called only once before permanently returning an error and
* **RetryCount**: that will invoke `AttemptMethod` for a maximum `RetryCount` number of times and each attempt must be separated by at least `RetryDelay` amount of time.

Also note that this `retrier` abstraction doesn’t perform the retries itself. By “performing the retry”, I mean “invoking the `AttemptMethod`. Instead, a caller must call the `retrier.TriggerRetry()` method.
* either a previous call to `TriggerRetry()` made a successful call to `AttemptMethod` and it will immediately return `OK`;
* either all previous calls to `AttemptMethod` failed and we have already reached the `RetryCount` number of attempts and `TriggerRetry()` will return `PermaFail`;
* either retries are still allowed. But depending on whether the last call to `AttemptMethod` happened more or less than `RetryDelay` ago, either `AttemptMethod` will be called by `TriggerRetry()`, either `TriggerRetry()` will immediately return `FailWillRetry` without calling `AttemptMethod`.

[`retrier` README](https://github.com/DataDog/datadog-agent/blob/master/pkg/util/retry/README.md)

Reply:
Thanks for the explanation 👌

Reply:
I am bit surprise that we need to handle the retry ourself, I would have hidden this in the implementation of the `Retrier` ..